### ISSUE TITLE
Automatically test which workflows are available

### ISSUE THREAD

**Issue by akreshuk** [ilastik/ilastik#1485]
Right now, we test for installation correctness by reading the list of workflows on the start-up screen. It would be nice if we had an installation test which would try to import them all and report which ones are available in the end. For bonus points, it could even report which solvers are available for the workflows with multiple back-ends.

**Comment by akreshuk** [comment]
To summarize our offline discussion with @chaubold , we should have 2 tests. The first one will check out a fresh dev environment from ilastik-build-conda on a machine with solvers, and then check for workflow availability and run all tests with different solvers/no solvers combinations. This we will run before tagging a release (or, for more fastidious developers, with any frequency up to after every commit). Unfortunately, because of solver licence issues, it can't be done on Circle CI. The second test will be run after a binary is built. It downloads a binary and checks which workflows are available, then adds and removes solvers one by one and checks that the right workflows with right options show up. 

@k-dominik , could you take a look at this for Linux?

**Comment by chaubold** [comment]
Thanks @akreshuk! To state a few more of the technical details how we thought this would happen:

* adding/removing solvers should mean nothing but placing/removing the respective `*.so` files into the `lib` folder, probably best done by our scripts in here: https://github.com/ilastik/ilastik/tree/master/scripts
* testing which workflows show up can be done by firing up ilastik and then parsing the `ilastik_log.txt` file, as ilastik spits out warnings for each workflow that could not be loaded

**Comment by k-dominik** [comment]
So, are you suggesting to setup a local machine with jenkins (or alike)?

**Comment by akreshuk** [comment]
That's an option too. I was, in the old-fashioned way, thinking of a script that every developer could run on his/her own machine, giving it the paths to cplex and gurobi. Perhaps that would introduce some interesting variability compared to all running on the same machine? But I'm not set on this, let's just start and figure out what's the easiest as we go. By "the easiest" I mean introduces the least friction in the development process.

**Comment by k-dominik** [comment]
@akreshuk I agree about the variability. I, however would be afraid that developer environments could be too polluted. And is any of the known developers even using CPLEX right now?

**Comment by akreshuk** [comment]
We all have cplex and you should get it too ;) or maybe not, so that someone can test the no-solvers case. For the pollution, I agree, that's why I'd rather have a script check out a fresh environment and do everything around it. This test environment can then be deleted. More concretely, I'm thinking of the following scenario:
- we think we are ready for a release with the current state of master in our repos
- a volunteer tester(s) checks out a conda environment ilastik-test-prerelease, which takes all the packages in ilastik-everything, but for those which come from our repos it replaces the tags with current master versions (or a given tag, that's fine too). 
- with this environment activated, we run all the tests with all solver combinations. 
- fix all errors, delete the environment once done, tag, package, do the binary tests. Go back to the first item.

The part that is not clear to me is how to embed the up-for-testing version of our repos into a conda environment without all the manual steps of deleting ilasitk-meta and replacing it with git versions. But it doesn't sound technically impossible.

**Comment by k-dominik** [comment]
> The part that is not clear to me is how to embed the up-for-testing version of our repos into a conda environment without all the manual steps of deleting ilasitk-meta and replacing it with git versions. But it doesn't sound technically impossible.

The current approach with replacing ilastik-meta has been used a lot and only consists of a handful lines of code. Is there any downside to it that I do not see?

**Comment by akreshuk** [comment]
Well, I want it to happen automagically for some list of packages (any python package can be in there). But you are right, nothing that a script couldn't do.
